napari/components/_dims/model.py

```diff
@@ -1,143 +1,297 @@
+            self.events.axis(axis=axis)
 
-        for layer in self.viewer.layers:
-            dims = layer.ndim
-            if dims > max_dims:
-                max_dims = dims
+    def set_mode(self, axis: int, mode: DimsMode):
+        """Sets the mode: POINT or INTERVAL
 
-        self._max_dims = max_dims
+        Parameters
+        ----------
+        axis : int
+            Dimension index
+        mode : POINT or INTERVAL
+            Whether dimension is in the POINT or INTERVAL mode
+        """
+        if self.mode[axis] != mode:
+            self.mode[axis] = mode
+            self.events.axis(axis=axis)
 
-        self._need_slider_update = True
-        self._update()
+    def set_display(self, axis: int, display: bool):
+        """Sets the display boolean flag for a given axis
 
-    def _calc_max_shape(self):
-        """Calculates the maximum shape of the contained images.
+        Parameters
+        ----------
+        axis : int
+            Dimension index
+        display : bool
+            Bool which is `True` for display and `False` for slice or project.
         """
-        shapes = (layer.shape for layer in self.viewer.layers)
-        self._max_shape = _compute_max_shape(shapes, self.max_dims)
+        if self.display[axis] != display:
+            self.display[axis] = display
+            self.events.axis(axis=axis)
 
-    def _on_layers_change(self, event):
-        """Called whenever a layer is changed.
+    def _set_2d_viewing(self):
+        """Sets the 2d viewing
         """
-        self._child_layer_changed = True
-        self._update()
+        self.display = [False] * len(self.display)
+        self.display[-1] = True
+        self.display[-2] = True
+
+    def _add_axes(self, axis: int, no_event=None):
+        """Makes sure that the given axis is in the dimension model
+
+        Parameters
+        ----------
+        axis : int
+            Dimension index
 
-    def _update(self):
-        """Updates the viewer.
+        Returns
+        -------
+        dimensions : list
+            List of axes
         """
-        if self._child_layer_changed:
-            self._child_layer_changed = False
-            self._recalc_max_dims = True
-            self._recalc_max_shape = True
-            self._need_slider_update = True
-
-        if self._need_redraw:
-            self._need_redraw = False
-            self.viewer._update_layers()
-
-        if self._recalc_max_dims:
-            self._recalc_max_dims = False
-            self._calc_max_dims()
-
-        if self._recalc_max_shape:
-            self._recalc_max_shape = False
-            self._calc_max_shape()
-
-        if self._need_slider_update:
-            self._need_slider_update = False
-            self._update_sliders()
+        if axis >= self.ndims:
+            old_ndims = self.ndims
+            margin_length = 1 + axis - self.ndims
+            self.range.extend([(0.0, 1.0, 0.01)] * (margin_length))
```

Comment:
This one is a bit weird if one can't override this behavior, but we can leave it be for now...

Reply:
agreed, should be settable, can fix later